hashjoin.go

```diff
@@ -0,0 +1,393 @@
+package exectoy
+
+// todo(changangela) hash table should be able to rehash or determine bucket size dynamically
+const hashTableBucketSize = 1 << 16
+
+type hashTableInt struct {
+	first []int
+	next []int
+
+	keys intColumn
+	values []intColumn
+
+	nOutCols int
+	count int
+	bucketSize int
+
+	// allocated represents the reserved number of keys that the hashTableInt
+	// can store in its current state.
+	allocated int
+}
+
+func makeHashTableInt(bucketSize int, nOutCols int) *hashTableInt {
+	values := make([]intColumn, nOutCols)
+	for i := 0; i < nOutCols; i++ {
+		values[i] = make(intColumn, 1)
+	}
+
+	hashTable := &hashTableInt{
+		first: make([]int, bucketSize),
+		next: make([]int, 1),
+		keys: make([]int, 1),
+		values: values,
+		nOutCols: nOutCols,
+		bucketSize: bucketSize,
+		allocated: 1,
+	}
+
+	return hashTable
+}
+
+func (hashTable *hashTableInt) hashInt(key int) int {
+	// todo(changangela) replace with a better hash function or use a library
+	return key % hashTable.bucketSize
+}
+
+func (hashTable *hashTableInt) grow(amount int) {
+	hashTable.next = append(hashTable.next, make([]int, amount)...)
+	hashTable.keys = append(hashTable.keys, make(intColumn, amount)...)
+
+	for i := 0; i < hashTable.nOutCols; i++ {
+		hashTable.values[i] = append(hashTable.values[i], make(intColumn, amount)...)
+	}
+	hashTable.allocated += amount
+}
+
+func (hashTable *hashTableInt) insert(hashKey int, key int) (id int) {
```

Comment:
This seems to be unused by the group implementation - why? Also, please add a comment on what it does.

Reply:
I wrote this function before implementing the second builder and it does a bit too much since it inserts to `keys` which we don't want to do if we are preloading everything.